README.md
# bucketstats

Data science tools for analyzing histograms using classic probbility theory and
statistics.  Useful when a dataset is too large to load into memory, but a
histogram is available.

[![Build Status](https://travis-ci.com/mprelee/bucketstats.svg?branch=master)](https://travis-ci.com/mprelee/bucketstats)


ChangeLog
CHANGES
=======

* Removes extra python versions; only 3.6

v0.3.3
------

* Removes duplicate py37 from travis yml

v0.3.2
------

* Update changelog and add skip\_cleanup to True

v0.3.1
------

* fix indentation
* Update changelog
* Adds api token for pypi
* Adds \_\_init\_\_.py and updates testing command in travis

v0.3
----

* Adds build badge to readme
* Updates changelog
* Adds verbose testing
* Removed cached\_property as it was blocking nosestests.  Fixes std dev and variance
* autopep8
* Comments out code for now
* Adds variance, std dev, entropy, mode, modes, and other basic stats.  Changes @property to @cached\_property for slight speedups
* Rewrites discrete distribution propertly with tests and comments
* Removes util.py for now
* Start refactoring to move everything into discrete\_distribution.py
* Changes test call
* Removes \_\_init\_\_,py.  Fixes package structure
* Adds DiscreteDistribution class and some checking of valid histograms

v0.2.4
------

* Try to fix issue by checking if a list
* Try to convert to list before extracting median values
* Adds a file to gitignore
* autopep8
* Updates travis to use nosetests
* More shuffling around.  Removed leading underscore so code could be shared
* Restructures import structure
* Fixes some import problems
* Adds \_safe\_getattr
* Adds travis ci

v0.2.3
------

* Try and fix setup.cfg again

v0.2.2
------

* changelog

v0.2.1
------

* Fixes setup.cfg

v0.2
----

* Updates tests
* Updates survival function estimators
* Adds util.py, containing a safe\_divide function
* Adds a test for rcumsum
* Add survival.py

v0.1.0
------

* Adds some basic distribution functions.  Adds \_\_init\_\_.py.  Adds test
* Initial commit
* Initial commit


test/test_util.py
#import doctest
#from bucketstats import util
#
#def load_tests(loader, tests, ignore):
#    tests.addTests(doctest.DocTestSuite(util))
#    return tests
#


test/test_basics.py
import doctest
from bucketstats import basics

def load_tests(loader, tests, ignore):
    tests.addTests(doctest.DocTestSuite(basics))
    return tests



test/test_probability.py
#import doctest
#from bucketstats import probability
#
#def load_tests(loader, tests, ignore):
#    tests.addTests(doctest.DocTestSuite(probability))
#    return tests
#


bucketstats/util.py
import numpy as np
import pandas as pd


def assert_valid_hist(hist: pd.Series):
    """Ensure histogram is numeric and sorted in increasing order."""
    assert isinstance(hist, pd.Series), "Expected pd.Series; got {}".format(type(hist))
    assert hist.index.is_numeric(), "Expected numeric Index; got {}".format(hist.index)
    assert hist.index.is_monotonic_increasing, "Expected monotonic increasing Index; got {}".format(str(hist.index))


def safe_divide(numer: pd.Series, denom: pd.Series, fill) -> pd.Series:
    """
    Safely divide two series by replacing zero denom with specified fill value

    >>> numer = pd.Series([1,1,1])
    >>> denom = pd.Series([0,1,2])
    >>> fill = 0
    >>> safe_divide(numer, denom, fill)
    0    0.0
    1    1.0
    2    0.5
    dtype: float64
    """

    result = numer / denom
    result[denom == 0] = fill
    return result


bucketstats/survival.py
# survival.py
# Matthew Prelee

__all__ = ['survival','kaplan_meier',]

import numpy as np
import pandas as pd
from bucketstats.basics import cmf, median, pmf, rcmf, rcumsum
from bucketstats.util import safe_divide


def survival(x: pd.Series) -> pd.Series:
    """Survival function.  Alias for reverse cumulative mass function.

    :param x: binned histogram.
    :type : pandas.Series
    :returns:  pandas.Series -- x normalized and then reverse cumulative sum.

    >>> x = pd.Series(data=[1,1,1,1,1], index=range(5))
    >>> survival(x)
    0    1.0
    1    0.8
    2    0.6
    3    0.4
    4    0.2
    dtype: float64

    >>> x = pd.Series(data=[5,6,7,8], index=range(4))
    >>> survival(x)
    0    1.000000
    1    0.807692
    2    0.576923
    3    0.307692
    dtype: float64
    """

    return rcmf(x)


def _ni(x: pd.Series) -> pd.Series:
    return rcumsum(x)


def _di(x: pd.Series, x_obs: pd.Series) -> pd.Series:
    return x_obs if x_obs is not None else x


def kaplan_meier(x: pd.Series, x_obs: pd.Series=None) -> pd.Series:
    """
    Kaplan-Meier estimate of survival function.
    https://en.wikipedia.org/wiki/Kaplan–Meier_estimator

    :param x: binned histogram.
    :type : pandas.Series
    :param x_obs: observed values.  If None, assumes x=x_obs
    :type : pandas.Series
    :returns:  pandas.Series -- Kaplan-Meier estimate of survival
    function of x

    >>> x = pd.Series(data=[1,1,1,1,1], index=range(5))
    >>> kaplan_meier(x)
    0    0.8
    1    0.6
    2    0.4
    3    0.2
    4    0.0
    dtype: float64
    """

    ni = _ni(x)
    di = _di(x, x_obs)
    return np.cumprod(1-safe_divide(di,ni,0))


def var_kaplan_meier(x: pd.Series, x_obs: pd.Series=None) -> pd.Series:
    """
    Kaplan-Meier estimate of survival function.
    https://en.wikipedia.org/wiki/Kaplan–Meier_estimator

    :param x: binned histogram.
    :type : pandas.Series
    :param x_obs: observed values.  If None, assumes x=x_obs
    :type : pandas.Series
    :returns:  pandas.Series -- Kaplan-Meier estimate of survival
    function of x

    >>> x = pd.Series(data=[1,1,1,1,1], index=range(5))
    >>> var_kaplan_meier(x)
    0    0.512000
    1    0.270000
    2    0.106667
    3    0.020000
    4    0.000000
    dtype: float64
    """

    ni, di = _ni(x), _di(x, x_obs)
    km_est = kaplan_meier(x,x_obs)
    inner = safe_divide(di, ni*(ni-di), 0)
    return km_est**2 * rcumsum(inner)


def nelson_aalen(x: pd.Series, x_obs: pd.Series=None) -> pd.Series:
    """
    Nelson-Aalen estimate of cumulative hazard rate.
    https://en.wikipedia.org/wiki/Nelson–Aalen_estimator

    :param x: binned histogram.
    :type : pandas.Series
    :param x_obs: observed values.  If None, assumes x=x_obs
    :type : pandas.Series
    :returns:  pandas.Series -- Kaplan-Meier estimate of survival
    function of x

    >>> x = pd.Series(data=[1,1,1,1,1], index=range(5))
    >>> nelson_aalen(x)
    0    2.283333
    1    2.083333
    2    1.833333
    3    1.500000
    4    1.000000
    dtype: float64

    """

    ni = _ni(x)
    di = _di(x, x_obs)
    return rcumsum(di/ni)


def var_nelson_aalen(x: pd.Series, x_obs: pd.Series=None) -> pd.Series:
    """
    Variance of Nelson-Aalen estimate of cumulative hazard rate.
    https://www.statsdirect.com/help/survival_analysis/kaplan.htm

    :param x: binned histogram.
    :type : pandas.Series
    :param x_obs: observed values.  If None, assumes x=x_obs
    :type : pandas.Series
    :returns:  pandas.Series -- Kaplan-Meier estimate of survival
    function of x

    >>> x = pd.Series(data=[1,1,1,1,1], index=range(5))
    >>> var_nelson_aalen(x)
    0    0.800000
    1    0.750000
    2    0.666667
    3    0.500000
    4    0.000000
    dtype: float64
    """

    numer = var_kaplan_meier(x, x_obs)
    denom = kaplan_meier(x, x_obs)**2
    return safe_divide(numer, denom, 0) 


if __name__ == '__main__':
    import doctest
    doctest.testmod()



bucketstats/probability.py
# distributions.py
# Matthew Prelee

import numpy as np
import pandas as pd

#def rcumsum(hist: pd.Series) -> pd.Series:
#    """Reverse cumulative sum of a data Series..
#
#    >>> hist = pd.Series(range(5))
#    >>> rcumsum(hist)
#    0    10
#    1    10
#    2     9
#    3     7
#    4     4
#    dtype: int64
#    """
#    pass
#
#
#def pmf(hist: pd.Series) -> pd.Series:
#    pass
#
#def cmf(hist: pd.Series) -> pd.Series:
#    """Cumulative mass function.
#
#    :param hist: binned histogram.
#    :type : pandas.Series
#    :returns:  pandas.Series -- hist normalized by sum(hist) and summed
#
#    >>> hist = pd.Series(data=[1,1,1,1,1], index=range(5))
#    >>> cmf(hist)
#    0    0.2
#    1    0.4
#    2    0.6
#    3    0.8
#    4    1.0
#    dtype: float64
#
#    >>> hist = pd.Series(data=[5,6,7,8], index=range(4))
#    >>> cmf(hist)
#    0    0.192308
#    1    0.423077
#    2    0.692308
#    3    1.000000
#    dtype: float64
#    """
#    pass
#
#
#def rcmf(hist: pd.Series) -> pd.Series:
#    """Reverse cumulative mass function.
#
#    :param hist: binned histogram.
#    :type : pandas.Series
#    :returns:  pandas.Series -- hist normalized and then reverse cumulative sum.
#
#    >>> hist = pd.Series(data=[1,1,1,1,1], index=range(5))
#    >>> rcmf(hist)
#    0    1.0
#    1    0.8
#    2    0.6
#    3    0.4
#    4    0.2
#    dtype: float64
#
#    >>> hist = pd.Series(data=[5,6,7,8], index=range(4))
#    >>> rcmf(hist)
#    0    1.000000
#    1    0.807692
#    2    0.576923
#    3    0.307692
#    dtype: float64
#
#    """
#    pass
#
#
#def median(hist: pd.Series) -> float:
#    """Median of weighted value counts.
#
#    :param hist: weighted value counts.
#    :type : pandas.Series
#    :returns:  float -- median of hist.index weighted by hist.values
#
#    >>> hist = pd.Series([1,1,1], index=range(3))
#    >>> median(hist)
#    1.0
#
#    >>> hist = pd.Series([9,1], index=range(2))
#    >>> median(hist)
#    0.0
#
#    >>> hist = pd.Series([1,1,1,1,1,1], index=range(6))
#    >>> median(hist)
#    2.5
#
#    >>> hist = pd.Series([3,4,3,4], index=range(4))
#    >>> median(hist)
#    1.5
#
#    """
#    pass
#
#
#def mean(hist: pd.Series) -> float:
#    pass


bucketstats/survival_analysis.py
# survival_analysis.py
# Matthew Prelee

#import numpy as np
#import pandas as pd
#from probability import *
#from util import safe_divide
#
#
#def survival(x: pd.Series) -> pd.Series:
#    """Survival function.  Alias for reverse cumulative mass function.
#
#    :param x: binned histogram.
#    :type : pandas.Series
#    :returns:  pandas.Series -- x normalized and then reverse cumulative sum.
#
#    >>> x = pd.Series(data=[1,1,1,1,1], index=range(5))
#    >>> survival(x)
#    0    1.0
#    1    0.8
#    2    0.6
#    3    0.4
#    4    0.2
#    dtype: float64
#
#    >>> x = pd.Series(data=[5,6,7,8], index=range(4))
#    >>> survival(x)
#    0    1.000000
#    1    0.807692
#    2    0.576923
#    3    0.307692
#    dtype: float64
#    """
#
#    return rcmf(x)
#
#
#def _ni(x: pd.Series) -> pd.Series:
#    return rcumsum(x)
#
#
#def _di(x: pd.Series, x_obs: pd.Series) -> pd.Series:
#    return x_obs if x_obs is not None else x
#
#
#def kaplan_meier(x: pd.Series, x_obs: pd.Series = None) -> pd.Series:
#    """
#    Kaplan-Meier estimate of survival function.
#    https://en.wikipedia.org/wiki/Kaplan–Meier_estimator
#
#    :param x: binned histogram.
#    :type : pandas.Series
#    :param x_obs: observed values.  If None, assumes x=x_obs
#    :type : pandas.Series
#    :returns:  pandas.Series -- Kaplan-Meier estimate of survival
#    function of x
#
#    >>> x = pd.Series(data=[1,1,1,1,1], index=range(5))
#    >>> kaplan_meier(x)
#    0    0.8
#    1    0.6
#    2    0.4
#    3    0.2
#    4    0.0
#    dtype: float64
#    """
#
#    ni = _ni(x)
#    di = _di(x, x_obs)
#    return np.cumprod(1 - safe_divide(di, ni, 0))
#
#
#def var_kaplan_meier(x: pd.Series, x_obs: pd.Series = None) -> pd.Series:
#    """
#    Kaplan-Meier estimate of survival function.
#    https://en.wikipedia.org/wiki/Kaplan–Meier_estimator
#
#    :param x: binned histogram.
#    :type : pandas.Series
#    :param x_obs: observed values.  If None, assumes x=x_obs
#    :type : pandas.Series
#    :returns:  pandas.Series -- Kaplan-Meier estimate of survival
#    function of x
#
#    >>> x = pd.Series(data=[1,1,1,1,1], index=range(5))
#    >>> var_kaplan_meier(x)
#    0    0.512000
#    1    0.270000
#    2    0.106667
#    3    0.020000
#    4    0.000000
#    dtype: float64
#    """
#
#    ni, di = _ni(x), _di(x, x_obs)
#    km_est = kaplan_meier(x, x_obs)
#    inner = safe_divide(di, ni * (ni - di), 0)
#    return km_est**2 * rcumsum(inner)
#
#
#def nelson_aalen(x: pd.Series, x_obs: pd.Series = None) -> pd.Series:
#    """
#    Nelson-Aalen estimate of cumulative hazard rate.
#    https://en.wikipedia.org/wiki/Nelson–Aalen_estimator
#
#    :param x: binned histogram.
#    :type : pandas.Series
#    :param x_obs: observed values.  If None, assumes x=x_obs
#    :type : pandas.Series
#    :returns:  pandas.Series -- Kaplan-Meier estimate of survival
#    function of x
#
#    >>> x = pd.Series(data=[1,1,1,1,1], index=range(5))
#    >>> nelson_aalen(x)
#    0    2.283333
#    1    2.083333
#    2    1.833333
#    3    1.500000
#    4    1.000000
#    dtype: float64
#
#    """
#
#    ni = _ni(x)
#    di = _di(x, x_obs)
#    return rcumsum(di / ni)
#
#
#def var_nelson_aalen(x: pd.Series, x_obs: pd.Series = None) -> pd.Series:
#    """
#    Variance of Nelson-Aalen estimate of cumulative hazard rate.
#    https://www.statsdirect.com/help/survival_analysis/kaplan.htm
#
#    :param x: binned histogram.
#    :type : pandas.Series
#    :param x_obs: observed values.  If None, assumes x=x_obs
#    :type : pandas.Series
#    :returns:  pandas.Series -- Kaplan-Meier estimate of survival
#    function of x
#
#    >>> x = pd.Series(data=[1,1,1,1,1], index=range(5))
#    >>> var_nelson_aalen(x)
#    0    0.800000
#    1    0.750000
#    2    0.666667
#    3    0.500000
#    4    0.000000
#    dtype: float64
#    """
#
#    numer = var_kaplan_meier(x, x_obs)
#    denom = kaplan_meier(x, x_obs)**2
#    return safe_divide(numer, denom, 0)
#
#
#if __name__ == '__main__':
#    import doctest
#    doctest.testmod()


bucketstats/discrete_distribution.py
# discrete_distribution.py

import pandas as pd
import numpy as np
from typing import List

# TODO: Replace with functools.property for py3.8
#from property import property


def _assert_valid_hist(hist: pd.Series):
    """Ensure histogram is numeric and sorted in increasing order.

    >>> a = "not_a_histogram"
    >>> DiscreteDistribution(a)
    Traceback (most recent call last):
    ...
    AssertionError: Expected pd.Series; got <class 'str'>

    >>> b = pd.Series(data=[1,2,3],index=['a','b','c'])
    >>> DiscreteDistribution(b)
    Traceback (most recent call last):
    ...
    AssertionError: Expected numeric Index; got Index(['a', 'b', 'c'], dtype='object')

     >>> c = pd.Series(data=[1,2,3],index=[4,6,5])
    >>> DiscreteDistribution(c)
    Traceback (most recent call last):
    ...
    AssertionError: Expected monotonic increasing Index; got Int64Index([4, 6, 5], dtype='int64')

    """
    assert isinstance(
        hist, pd.Series), "Expected pd.Series; got {}".format(
        type(hist))
    assert hist.index.is_numeric(), "Expected numeric Index; got {}".format(hist.index)
    assert hist.index.is_monotonic_increasing, "Expected monotonic increasing Index; got {}".format(
        str(hist.index))


def _safe_divide(numer: pd.Series, denom: pd.Series, fill) -> pd.Series:
    """
    Safely divide two series by replacing zero denom with specified fill value

    >>> numer = pd.Series([1,1,1])
    >>> denom = pd.Series([0,1,2])
    >>> fill = 0
    >>> _safe_divide(numer, denom, fill)
    0    0.0
    1    1.0
    2    0.5
    dtype: float64
    """

    result = numer / denom
    result[denom == 0] = fill
    return result


class DiscreteDistribution:
    """A class for enabling probabily analysis on finite discrete distributions
    resulting from a histogram.  Histograms are easily generated in libraries
    like numpy and pandas, or by running a simple SQL query against a
    database.  This class transforms the histogram into various classic
    probability distributions and other properties.
    """

    def __init__(self, hist):
        """Constructor for  DiscreteDistribution object.
        The provided histogram must be a pandas Series
        whose Index must be numeric and monotonically increasing
        (and therefore, sorted)

        >>> hist = pd.Series(range(5))
        >>> d = DiscreteDistribution(hist)
        """
        _assert_valid_hist(hist)
        self._hist = hist

    def toseries(self) -> pd.Series:
        """ Return distribution as a series.  Returns original histogram

        >>> hist = pd.Series(range(5))
        >>> d = DiscreteDistribution(hist)
        >>> d.toseries()
        0    0
        1    1
        2    2
        3    3
        4    4
        dtype: int64
        """
        return self._hist

    @property
    def index(self) -> pd.Index:
        """Support of underlying histogram, as a pandas Index.

        >>> hist = pd.Series(range(5))
        >>> d = DiscreteDistribution(hist)
        >>> d.index
        RangeIndex(start=0, stop=5, step=1)
        """

        return self._hist.index

    @property
    def values(self):
        """Values of underlying histogram, which are the counts
        of each histogram "bucket" of the index
        >>> hist = pd.Series(range(5))
        >>> d = DiscreteDistribution(hist)
        >>> d.values
        array([0, 1, 2, 3, 4])
        """

        return self._hist.values

    @property
    def support(self):
        """ Nonzero support of underlying distribution """
        return self.index[self.values > 0]

    @property
    def sum(self):
        """Sum of underlying histogram values

        >>> hist = pd.Series(range(5))
        >>> d = DiscreteDistribution(hist)
        >>> d.sum
        10
        """

        return self._hist.sum()

    @property
    def cumsum(self) -> pd.Series:
        """Cumulative sum of the underlying histogram.
        >>> hist = pd.Series(range(5))
        >>> d = DiscreteDistribution(hist)
        >>> d.sum
        10
        """
        return np.cumsum(self._hist)

    @property
    def rcumsum(self) -> pd.Series:
        """Reverse cumulative sum of the underlying histogram.
        >>> hist = pd.Series(range(5))
        >>> d = DiscreteDistribution(hist)
        >>> d.rcumsum
        0    10
        1    10
        2     9
        3     7
        4     4
        dtype: int64
        """

        return np.cumsum(self._hist[::-1])[::-1]

    @property
    def cmf(self) -> pd.Series:
        """Cumulative mass function.

        >>> hist = pd.Series(data=[1,1,1,1,1], index=range(5))
        >>> d = DiscreteDistribution(hist)
        >>> d.cmf
        0    0.2
        1    0.4
        2    0.6
        3    0.8
        4    1.0
        dtype: float64

        >>> hist = pd.Series(data=[5,6,7,8], index=range(4))
        >>> d = DiscreteDistribution(hist)
        >>> d.cmf
        0    0.192308
        1    0.423077
        2    0.692308
        3    1.000000
        dtype: float64
        """
        return self.cumsum / self.sum

    @property
    def rcmf(self) -> pd.Series:
        """Reverse cumulative mass function.

        >>> hist = pd.Series(data=[1,1,1,1,1], index=range(5))
        >>> d = DiscreteDistribution(hist)
        >>> d.rcmf
        0    1.0
        1    0.8
        2    0.6
        3    0.4
        4    0.2
        dtype: float64

        >>> hist = pd.Series(data=[5,6,7,8], index=range(4))
        >>> d = DiscreteDistribution(hist)
        >>> d.rcmf
        0    1.000000
        1    0.807692
        2    0.576923
        3    0.307692
        dtype: float64
        """

        return self.rcumsum / self.sum

    @property
    def pmf(self) -> pd.Series:
        """Probability mass function.

        >>> hist = pd.Series(data=[1,1,1,1,1], index=range(5))
        >>> d = DiscreteDistribution(hist)
        >>> d.pmf
        0    0.2
        1    0.2
        2    0.2
        3    0.2
        4    0.2
        dtype: float64
        """
        return self._hist / self.sum

    @property
    def mean(self):
        """Expected value of distribution
        >>> hist = pd.Series(range(5))
        >>> d = DiscreteDistribution(hist)
        >>> d.mean
        3.0
        """
        return np.inner(self.index, self.pmf)

    @property
    def median(self) -> float:
        """Median of underlying distribution

        >>> hist = pd.Series([1,1,1], index=range(3))
        >>> d = DiscreteDistribution(hist)
        >>> d.median
        1.0

        >>> hist = pd.Series([9,1], index=range(2))
        >>> d = DiscreteDistribution(hist)
        >>> d.median
        0.0

        >>> hist = pd.Series([1,1,1,1,1,1], index=range(6))
        >>> d = DiscreteDistribution(hist)
        >>> d.median
        2.5

        >>> hist = pd.Series([3,4,3,4], index=range(4))
        >>> d = DiscreteDistribution(hist)
        >>> d.median
        1.5
        """

        tmp = 2 * self.cumsum - self.sum
        _idx_left = np.searchsorted(tmp, 0, side='left').tolist()
        _idx_right = np.searchsorted(tmp, 0, side='right').tolist()
        idx_left = _idx_left if not isinstance(
            _idx_left, list) else _idx_left[0]
        idx_right = _idx_right if not isinstance(
            _idx_right, list) else _idx_right[0]
        if idx_left == idx_right:
            return float(idx_left)
        else:
            return (idx_left + idx_right) / 2

    @property
    def modes(self) -> List:
        """Modes of underlying distribution as a list

        >>> hist = pd.Series([1,1,1,4], index=range(4))
        >>> d = DiscreteDistribution(hist)
        >>> d.modes
        [3]

        >>> hist = pd.Series([3,4,3,4], index=range(4))
        >>> d = DiscreteDistribution(hist)
        >>> d.modes
        [1, 3]
        """
        highest_frequency = max(self.values)
        return self.index[self.values == highest_frequency].tolist()

    @property
    def mode(self):
        """Mode of underlying unimodal distribution.  Throws an
        exception if distribution is multimodal.

        >>> hist = pd.Series([1,1,1,4], index=range(4))
        >>> d = DiscreteDistribution(hist)
        >>> d.mode
        3

        >>> hist = pd.Series([3,4,3,4], index=range(4))
        >>> d = DiscreteDistribution(hist)
        >>> d.mode
        Traceback (most recent call last):
        ...
        AssertionError: Distribution is multimodal with modes [1, 3]
        """
        assert(len(self.modes) == 1), "Distribution is multimodal with modes {}".format(
            self.modes)
        return self.modes[0]

    @property
    def variance(self):
        """Variance of distribution
        >>> hist = pd.Series(data=[10,10], index=[0,1])
        >>> d = DiscreteDistribution(hist)
        >>> d.variance
        0.25
        """
        return np.inner(self.pmf, np.power(self.index, 2)) - self.mean**2

    @property
    def std_dev(self):
        """Standard deviation of distribution
        >>> hist = pd.Series(data=[10,10], index=[0,1])
        >>> d = DiscreteDistribution(hist)
        >>> d.std_dev
        0.5
        """
        return np.sqrt(self.variance)

    @property
    def entropy(self) -> float:
        """Shannon entropy of this distribution
        >>> hist = pd.Series(data=[10,10], index=[0,1])
        >>> d = DiscreteDistribution(hist)
        >>> d.entropy
        1.0
        """
        return np.sum(-self.pmf * np.log2(self.pmf))

    # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    # Survival Analysis
    # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


if __name__ == '__main__':
    import doctest
    doctest.testmod()


bucketstats/basics.py
# distributions.py
# Matthew Prelee

__all__ = ['rcumsum', 'pmf', 'cmf', 'rcmf', 'median']

import numpy as np
import pandas as pd
from util import safe_getitem


def rcumsum(x: pd.Series) -> pd.Series:
    """Reverse cumulative sum of a data Series..

    :param x: binned histogram.
    :type : pandas.Series
    :returns:  pandas.Series -- reverse cumulative distribution

    >>> x = pd.Series(range(5))
    >>> rcumsum(x)
    0    10
    1    10
    2     9
    3     7
    4     4
    dtype: int64
    """

    return np.cumsum(x[::-1])[::-1]


def pmf(x: pd.Series) -> pd.Series:
    """Probability mass function.

    :param x: binned histogram.
    :type : pandas.Series
    :returns:  pandas.Series -- x normalized by sum(x)

    >>> x = pd.Series(data=[1,1,1,1,1], index=range(5))
    >>> pmf(x)
    0    0.2
    1    0.2
    2    0.2
    3    0.2
    4    0.2
    dtype: float64
    """

    return x/np.sum(x)


def cmf(x: pd.Series) -> pd.Series:
    """Cumulative mass function.

    :param x: binned histogram.
    :type : pandas.Series
    :returns:  pandas.Series -- x normalized by sum(x) and summed

    >>> x = pd.Series(data=[1,1,1,1,1], index=range(5))
    >>> cmf(x)
    0    0.2
    1    0.4
    2    0.6
    3    0.8
    4    1.0
    dtype: float64

    >>> x = pd.Series(data=[5,6,7,8], index=range(4))
    >>> cmf(x)
    0    0.192308
    1    0.423077
    2    0.692308
    3    1.000000
    dtype: float64
    """

    return np.cumsum(pmf(x))


def rcmf(x: pd.Series) -> pd.Series:
    """Reverse cumulative mass function.

    :param x: binned histogram.
    :type : pandas.Series
    :returns:  pandas.Series -- x normalized and then reverse cumulative sum.

    >>> x = pd.Series(data=[1,1,1,1,1], index=range(5))
    >>> rcmf(x)
    0    1.0
    1    0.8
    2    0.6
    3    0.4
    4    0.2
    dtype: float64

    >>> x = pd.Series(data=[5,6,7,8], index=range(4))
    >>> rcmf(x)
    0    1.000000
    1    0.807692
    2    0.576923
    3    0.307692
    dtype: float64

    """

    return rcumsum(pmf(x))


def median(x: pd.Series) -> float:
    """Median of weighted value counts.

    :param x: weighted value counts.
    :type : pandas.Series
    :returns:  float -- median of x.index weighted by x.values

    >>> x = pd.Series([1,1,1], index=range(3))
    >>> median(x)
    1.0

    >>> x = pd.Series([9,1], index=range(2))
    >>> median(x)
    0.0

    >>> x = pd.Series([1,1,1,1,1,1], index=range(6))
    >>> median(x)
    2.5

    >>> x = pd.Series([3,4,3,4], index=range(4))
    >>> median(x)
    1.5

    """
    tmp = 2*np.cumsum(x) - sum(x)
    idx_left  = safe_getitem(np.searchsorted(tmp, 0, side='left'),0)
    idx_right = safe_getitem(np.searchsorted(tmp, 0, side='right'),0)
    if idx_left == idx_right:
        return float(idx_left)
    else:
        return (idx_left + idx_right)/2


if __name__ == '__main__':
    import doctest
    doctest.testmod()

